### PR TITLE
[9.x] Allow enum route bindings to have default values

### PR DESCRIPTION
**Pull request by levu42**
It's possible already to bind route parameters to enums:

```php
enum Status: string {
  case Published = 'published';
  case Draft = 'draft';
}

Route::get('/posts/{status}', function(Status $status) {});
```

However, it's not yet possible to have default values for optional parameters, as I discovered in #44249:

```php
Route::get('/posts/{status?}', function(Status $status = Status::Published) {});
```

As discussed in #44249, I made the PR in a way, that it only applies, if the parameter is an enum, so it doesn't interfere with model binding in any way.

I'm not fully happy with the nested ternary operator, making it a bit more complex - I'm happy to change it so some other structure if you think something else is more readable.

I tried to keep the code change minimal to support this while having extensive tests, for them I looked at the existing backed enum binding tests and extended those.

### CODE REVIEW

File: src/Illuminate/Routing/RouteDependencyResolverTrait.php [hunk 1]
```diff
@@ -77,7 +78,11 @@ protected function transformDependency(ReflectionParameter $parameter, $paramete
         // the list of parameters. If it is we will just skip it as it is probably a model
         // binding and we do not want to mess with those; otherwise, we resolve it here.
         if ($className && ! $this->alreadyInParameters($className, $parameters)) {
-            return $parameter->isDefaultValueAvailable() ? null : $this->container->make($className);
+            $isEnum = method_exists(ReflectionClass::class, 'isEnum') ? (new ReflectionClass($className))->isEnum() : false;
+
```
Comment (BrandonSurowiec) [inline src/Illuminate/Routing/RouteDependencyResolverTrait.php:82]: You could tighten it up with:

```php
$isEnum = method_exists(ReflectionClass::class, 'isEnum') && (new ReflectionClass($className))->isEnum();
```

I think your nested ternary is understandable as-is.

Comment (levu42) [reply src/Illuminate/Routing/RouteDependencyResolverTrait.php:82]: @BrandonSurowiec Thanks for the idea, just changed it!

**Review by BrandonSurowiec — COMMENTED** [review]
Looks good to me!

**Comment by taylorotwell** [comment]
Thanks 👍

**Pull request merged**